src/main/java/com/vunet/keycloak/auth/otp/OtpAuthenticatorForm.java

```diff
@@ -0,0 +1,281 @@
+            form.setAttribute(LoginFormsProvider.USERNAME_HIDDEN, true);
+            form.setAttribute(LoginFormsProvider.REGISTRATION_DISABLED, true);
+            context.getAuthenticationSession().setAuthNote(USER_SET_BEFORE_USERNAME_PASSWORD_AUTH, "true");
+        } else {
+            context.getAuthenticationSession().removeAuthNote(USER_SET_BEFORE_USERNAME_PASSWORD_AUTH);
+            if (loginHint != null || rememberMeUsername != null) {
+                if (loginHint != null) {
+                    formData.add(AuthenticationManager.FORM_USERNAME, loginHint);
+                } else {
+                    formData.add(AuthenticationManager.FORM_USERNAME, rememberMeUsername);
+                    formData.add("rememberMe", "on");
+                }
+            }
+        }
+        Response challengeResponse = lchallenge(context, formData);
+        context.challenge(challengeResponse);
+    }
+
+    protected Response lchallenge(AuthenticationFlowContext context, MultivaluedMap<String, String> formData) {
+        LoginFormsProvider forms = context.form();
+
+        if (formData.size() > 0) forms.setFormData(formData);
+
+        return forms.createLoginUsernamePassword();
+    }
+
+
+    @Override
+    public void action(AuthenticationFlowContext context) {
+    	logger.info("User action made");
+        MultivaluedMap<String, String> formData = context.getHttpRequest().getDecodedFormParameters();
+        if (formData.containsKey("cancel")) {
+        	logger.info("canceling login");
+            context.cancelLogin();
+            return;
+        }
+        if (!validateForm(context, formData)) {
+            return;
+        }
+        context.success();
+    }
+    protected boolean validateForm(AuthenticationFlowContext context, MultivaluedMap<String, String> formData) {
+    	logger.info("validating credentials "+formData.toString());
+    	boolean result = validateUserAndPassword(context, formData);
+        if(result) {
+        	result= validateOTP(context,formData);
+        }
+        return result;
+
+    }
+        
+    public boolean validateOTP(AuthenticationFlowContext context, MultivaluedMap<String, String> inputData) {
+
+    	logger.info("validating otp");
+        String otp = inputData.getFirst(OTP);
+
+        UserModel userModel = context.getUser();
+        if (!enabledUser(context, userModel)) {
+            return false;
+        }
+
+        if (otp.isBlank()) {
+        	logger.info("otp is empty");
+        	 Response challengeResponse = challenge(context, Messages.MISSING_TOTP);
+             context.forceChallenge(challengeResponse);
+            return false;
+        }
+        boolean valid = mockvalidateWithNPCI(context, inputData);
+        if(!valid) {
+        	context.resetFlow();
+        }
+        
+		return valid;
+    }
+    
+    private boolean mockvalidateWithNPCI(AuthenticationFlowContext context, MultivaluedMap<String, String> inputData) {
```

Comment:
This looks like a mock function that you might've used for testing. Can you check if this is needed?

Reply:
This is not needed. Forgot to remove
Will remove this

Reply:
Thanks. Update it here once its done

Reply:
Have removed it